### ISSUE TITLE
Extension causes high cpu load when using modules

### ISSUE THREAD

**Issue by lunglin** [microsoft/vscode-go#2391]
[ms-vscode.Go-unresponsive.cpuprofile.txt](https://github.com/Microsoft/vscode-go/files/2986791/ms-vscode.Go-unresponsive.cpuprofile.txt)
- Issue Type: `Performance`
- Extension Name: `Go`
- Extension Version: `0.9.2`
- OS Version: `Linux x64 3.10.0-957.10.1.el7.x86_64`
- VSCode version: `1.32.3`

:warning: Make sure to **attach** this file from your *home*-directory: `/home/hlin/ms-vscode.Go-unresponsive.cpuprofile.txt` :warning:

Find more details here: https://github.com/Microsoft/vscode/wiki/Explain:-extension-causes-high-cpu-load

**Comment by alexanderbez** [comment]
I'm experiencing the same issue. After upgrading to go 1.12 and vscode 1.32.3, CPU usage is extremely high. Code completion and similar features take a long time to appear or work.

**Comment by gillesfabio** [comment]
Same issue too.

**Comment by ramya-rao-a** [comment]
Please run the `Developer: Open Process Explorer` command to see which process is causing the high cpu load and share a screen shot of the process explorer.

**Comment by ramya-rao-a** [comment]
If you are using Go modules, try using the new [language server from Google](https://github.com/golang/go/wiki/gopls)

**Comment by alexanderbez** [comment]
> Please run the `Developer: Open Process Explorer` command to see which process is causing the high cpu load and share a screen shot of the process explorer.

So I ran the process explorer, but unfortunately it did not show me any detailed processes and the CPU consumption was very off (only 3-5%). However, my other system resource utilization tools showed that the process was `gocode-gomod` and it was using 300+% CPU.

**Comment by alexanderbez** [comment]
The issue was resolved for me by following the suggestions listed by the link provided by @ramya-rao-a:

```
"go.useLanguageServer": true,
"go.alternateTools": {
     "go-langserver": "gopls"
},
"go.languageServerExperimentalFeatures": {
    "format": true,
    "autoComplete": true,
    "goToDefinition": true,
    "hover": true,
    "signatureHelp": true,
    "goToTypeDefinition": true
},
"go.buildOnSave": "off",
"go.vetOnSave": "off",
"editor.formatOnSave": true,
"editor.codeActionsOnSave": {
    "source.organizeImports": true
},
```

**Comment by ramya-rao-a** [comment]
`gocode-gomod` is the tool used for auto-completion if you are using modules.
I am soon going to have this extension use the go language server by default when using modules, so this problem should go away.

Thanks for the confirmation @alexanderbez 

@lunglin , @gillesfabio are you using modules as well?

**Comment by gillesfabio** [comment]
@ramya-rao-a thank you very much for your help. Yes, I'm using modules.

**Comment by ramya-rao-a** [comment]
@gillesfabio In that case, can you follow the suggestion in 
https://github.com/Microsoft/vscode-go/issues/2391#issuecomment-475839123 and try the Go language server as well and let me know if that eases things up?

**Comment by alexanderbez** [comment]
@ramya-rao-a just a quick update. I'm not totally sure if this is in the domain of vscode or the language server (gopls), but I've noticed that when linting or build errors occur and are resolved, the warning/error still persists.

**Comment by ramya-rao-a** [comment]
@alexanderbez That is being tracked in gopls at https://github.com/golang/go/issues/30786
Please subscribe to that issue.
The next update to the Go extension will have an option to opt out of the error features from the language feature, so you can then use that until the issue is fixed in gopls

**Comment by alexanderbez** [comment]
Thanks for the issue reference @ramya-rao-a. I'm a bit confused by your suggested fix in the meantime though? What setting(s) should I disable/enable?

**Comment by ramya-rao-a** [comment]
In the meantime, unfortunately there is no workaround if you are using `gopls`. But I will post an update here (mostly in the next day or 2) with a pointer to use our beta extension which will have the setting to opt-out of diagnostics from `gopls`

**Comment by lunglin** [comment]
@ramya-rao-a thank you very much for your help. Yes, I'm using modules,too.

**Comment by ramya-rao-a** [comment]
Thanks @lunglin 
I'll close this issue as the alternative to use `gopls` should fix all the modules related high cpu issues being discussed here.

In the next update, I will be adding a prompt to use `gopls` when we detect that the user is using Go modules

**Comment by aaronhnatiw** [comment]
I just wanted to let everyone know what worked for me, in case it helps others (and you as well, @ramya-rao-a, if you're designing docs around it):

I had to manually delete the previous gocode binary (found at `$~/go/bin/gocode` for me) before it effectively used `gopls` instead. This was done after closing vscode and before starting it up again.

Hope this helps.

**Comment by ramya-rao-a** [comment]
@aaronhnatiw Did you get to install `gopls` and reload VS Code window before attempting to delete `gocode`?

Until VS Code window is reloaded after installing `gopls`, the extension will not begin to use the language server

**Comment by aaronhnatiw** [comment]
You know what? I may have not properly installed `gopls` before, so please ignore my comment. It appears to be working now (I haven't seen the same `gocode` CPU issue pop up again once I installed `gopls` and reloaded VS Code).

For posterity's sake, I did the following (referencing the [official gopls wiki page](https://github.com/golang/go/wiki/gopls), who's specific instructions may change in the future):
1. installed `gopls`: `$ go get -u golang.org/x/tools/cmd/gopls`.
2. Added the following to my `settings.json` file in VS Code:
```json
"go.useLanguageServer": true,
"go.alternateTools": {
     "go-langserver": "gopls"
},
"go.languageServerExperimentalFeatures": {
    "format": true,
    "autoComplete": true,
    "diagnostics": true,
    "goToDefinition": true,
    "hover": true,
    "signatureHelp": true,
    "goToTypeDefinition": true
},
"go.buildOnSave": "off",
"go.vetOnSave": "off",
"[go]": {
    "editor.snippetSuggestions": "none",
    "editor.formatOnSave": true,
    "editor.codeActionsOnSave": {
        "source.organizeImports": true
    },
}
```
3. Reloaded VC Code.

Also note, for those who may be tempted to set `"go.buildOnSave": "on"` and `"go.vetOnSave": "on"` - these are set to `false`, because `gopls` would otherwise run its own diagnostics _on top of_ the diagnostics tools provided by the vscode-go extension (thus duplicating work unnecessarily and just adding more overhead).

Hope this helps provide a clear description of the workaround until the next version of vscode-go comes out with the fix.

**Comment by aaronhnatiw** [comment]
And it looks like I spoke too soon, unfortunately. Gocode is back using all the CPU it can take:

<img width="637" alt="gocode" src="https://user-images.githubusercontent.com/2362857/56146779-da033880-5f74-11e9-89a0-a2ac0ad568bd.png">

**Comment by ramya-rao-a** [comment]
@aaronhnatiw That is very weird, what version of the Go extension are you using? Can you try the [latest beta version](https://github.com/Microsoft/vscode-go/wiki/Use-the-beta-version-of-the-latest-Go-extension)?

Can you also check where all your above settings are made? In `User` or `Workspace` settings?

The settings done in `Workspace` settings override the ones in `User` and apply only to that workspace.

**Comment by alexanderbez** [comment]
Note, I have been using vscode with general success after switching to modules, but I consistently get the following which makes development pretty difficult:

```
The go-langserver server crashed 5 times in the last 3 minutes. The server will not be restarted.
```

Are there some logs I can search for to post here?

**Comment by ramya-rao-a** [comment]
@alexandrevez Can you switch to using `gopls` instead of `go-langserver`? The latter is not in active development anymore.
- Install the [beta version](https://github.com/Microsoft/vscode-go/wiki/Use-the-beta-version-of-the-latest-Go-extension)
- Reload your project. If you are using Go modules, you will get prompted to install and use `gopls`, go ahead with that
- Reload your window another time to activate the language server

**Comment by alexanderbez** [comment]
@ramya-rao-a I am using the beta version of the Go extension. This is my config:

```
    "go.useLanguageServer": true,
    "go.alternateTools": {
        "go-langserver": "gopls"
    },
    "go.languageServerExperimentalFeatures": {
        "format": true,
        "autoComplete": true,
        "goToDefinition": true,
        "hover": true,
        "signatureHelp": true,
        "goToTypeDefinition": true,
    },
    "go.testEnvVars": {
        "GO111MODULE": "on"
    },
    "go.buildOnSave": "off",
    "go.vetOnSave": "off",
    "editor.codeActionsOnSave": {
        "source.organizeImports": true
    },
    "go.lintTool": "golangci-lint",
    "go.lintFlags": [
        "--fast",
    ],
```

What do I need to do to get rid of `go-langserver` and use `gopls` instead? Note, I already have `gopls` installed.

**Comment by alexanderbez** [comment]
I removed `gopls` and `gocode` and restarte vscode, and I got `Extension causes high cpu load`. I've attached the log file.

[ms-vscode.Go-unresponsive.cpuprofile.txt](https://github.com/Microsoft/vscode-go/files/3090852/ms-vscode.Go-unresponsive.cpuprofile.txt)

**Comment by lunglin** [comment]
@ramya-rao-a I am using the beta version of the Go extension and my workspace configuration:
{
	"folders": [
		{
			"path": "go"
		}
	],
	"settings": {    
		"go.useLanguageServer": true,
		"go.alternateTools": {
			"go-langserver": "gopls"
		},
		"go.buildOnSave": "off",
		"go.vetOnSave": "off",
		"go.languageServerExperimentalFeatures": {
			"format": true,
			"autoComplete": true,
			"diagnostics": true,
			"goToDefinition": true,
			"hover": true,
			"signatureHelp": true,
			"goToTypeDefinition": true
		},
		"[go]": {
			"editor.snippetSuggestions": "none",
			"editor.insertSpaces": false,
			"editor.formatOnSave": true,
			"editor.codeActionsOnSave": {
				"source.organizeImports": true
			}
		}
	}
}

![image](https://user-images.githubusercontent.com/3055767/56334912-ea155500-61cc-11e9-9764-c9b431671585.png)
![image](https://user-images.githubusercontent.com/3055767/56335301-8855ea80-61ce-11e9-86bc-c64c6a9678c4.png)

**Comment by ramya-rao-a** [comment]
Apologies, the issue got locked unintentionally

@alexanderbez You can remove the below from your settings
```
"go.alternateTools": {
        "go-langserver": "gopls"
    }
"go.languageServerExperimentalFeatures": {
        "format": true,
        "autoComplete": true,
        "goToDefinition": true,
        "hover": true,
        "signatureHelp": true,
        "goToTypeDefinition": true,
    }
````

Since `gopls` is now the default language server to use, you don't need the alternateTools setting above
The default value of `go.languageServerExperimentalFeatures` is such that all features are enabled which is what you want based on what you have set.

Regarding high cpu usage, please run `Developer: Open Process Explorer` the next time to see which process is causing it. If it is indeed `gopls` then the next step is to log an issue for it.

**Comment by ramya-rao-a** [comment]
Closing this issue as it hasn't seen a response in a while.
If anyone is still facing this problem, please log a new issue.
